airbyte-integrations/connectors/source-mysql-v2/src/main/kotlin/io/airbyte/integrations/source/mysql/MysqlJdbcPartitionFactory.kt

```diff
@@ -0,0 +1,163 @@
+        } else {
+            if (sv.stateType != "cursor_based") {
+                // Loading value from catalog. Note there could be unexpected behaviors if user
+                // updates their schema but did not reset their state.
+                val pkChosenFromCatalog: List<Field> = stream.configuredPrimaryKey ?: listOf()
+                val pkLowerBound: JsonNode = Jsons.valueToTree(sv.pkValue)
+                val cursorChosenFromCatalog: Field =
+                    stream.configuredCursor as? Field ?: throw ConfigErrorException("no cursor")
+
+                // in a state where it's still in primary_key read part.
+                return MysqlJdbcSplittableSnapshotWithCursorPartition(
+                    selectQueryGenerator,
+                    streamState,
+                    pkChosenFromCatalog,
+                    lowerBound = listOf(pkLowerBound),
+                    upperBound = null,
+                    cursorChosenFromCatalog,
+                    cursorUpperBound = null,
+                )
+            }
+            // resume back to cursor based increment.
+            val cursor: Field = stream.fields.find { it.id == sv.cursorField.first() } as Field
+            var cursorCheckpoint: JsonNode =
+                when (cursor.type.airbyteType) {
+                    is LeafAirbyteType ->
+                        when (cursor.type.airbyteType as LeafAirbyteType) {
+                            LeafAirbyteType.INTEGER -> {
+                                Jsons.valueToTree(sv.cursors.toInt())
+                            }
+                            LeafAirbyteType.NUMBER -> {
+                                Jsons.valueToTree(sv.cursors.toDouble())
+                            }
+                            else -> Jsons.valueToTree(sv.cursors)
+                        }
+                    else ->
+                        throw IllegalStateException(
+                            "Cursor field must be leaf type but is ${cursor.type.airbyteType}."
+                        )
+                }
+            // Compose a jsonnode of cursor label to cursor value to fit in
+            // DefaultJdbcCursorIncrementalPartition
+            if (cursorCheckpoint == streamState.cursorUpperBound) {
+                // Incremental complete.
+                return null
+            }
+            return MysqlJdbcCursorIncrementalPartition(
+                selectQueryGenerator,
+                streamState,
+                cursor,
+                cursorLowerBound = cursorCheckpoint,
+                isLowerBoundIncluded = false,
+                cursorUpperBound = streamState.cursorUpperBound,
+            )
+        }
+    }
+
+    override fun split(
+        unsplitPartition: MysqlJdbcPartition,
+        opaqueStateValues: List<OpaqueStateValue>
+    ): List<MysqlJdbcPartition> {
+        // At this moment we don't support split on within mysql stream in any mode.
+        return listOf(unsplitPartition)
+    }
+}
```

Comment:
this all looks fine at first glance but I'm not looking at it in too much detail -- either it works or it doesn't
this needs to be determined by tests
as discussed, these can come in a subsequent PR

Reply:
yeah I think we need a test for factory at least and I agree it should be in the next pr